File: services/identite/src/identite/service.py
"""Service identite : ses routes sous /api/identite. Il ne parle jamais FHIR et n'est pas sur le réseau du noyau.

Il connecte les agents, les officines et les citoyens, et pose les deux cookies de leur session sur le
sous-domaine de l'application où ils se connectent (ADR 0004) : `__Host-session`, le jeton, et
`__Host-renouvellement`, l'identifiant de la session qui le renouvelle. Il émet les codes carnet
(ADR 0005). Sa base, ses règles et son guichet font le reste ; les routes n'y ajoutent que HTTP.
"""

import logging
import os
from collections.abc import AsyncIterator
from contextlib import asynccontextmanager
from typing import Annotated

from fastapi import APIRouter, Cookie, Depends, FastAPI, Form, HTTPException, Request, Response, status
from fastapi.responses import JSONResponse, RedirectResponse
from pydantic import BaseModel, StringConstraints

from commun.jeton import COOKIE_DE_SESSION, Agent, Porteur
from commun.service import creer_service
from identite import demonstration
from identite.base import BaseDIdentite
from identite.demonstration import CitoyenDeDemonstration, CompteDeDemonstration
from identite.guichet import Guichet, Refus, SessionAgent, SessionCitoyen, SessionOfficine, SessionOuverte
from identite.jetons import SignataireDeJetons
from identite.regles.applications import Application, application_de_l_hote, origine
from identite.regles.code_carnet import NPI, ROLES_QUI_EMETTENT_UN_CODE

SERVICE = "identite"
COOKIE_DE_RENOUVELLEMENT = "__Host-renouvellement"

logging.basicConfig(level=logging.WARNING, format="%(name)s : %(message)s")
logging.getLogger(SERVICE).setLevel(logging.INFO)

DOMAINE = os.environ.get("LAFIA_DOMAINE", "")
# La clé privée est lue au démarrage : sans elle, hors de localhost, le service ne démarre pas.
signataire = SignataireDeJetons.depuis_environnement()
jetons = signataire.verificateur()
soignant_connecte = jetons.agent(ROLES_QUI_EMETTENT_UN_CODE)


@asynccontextmanager
async def cycle_de_vie(app: FastAPI) -> AsyncIterator[None]:
    """Ouvre la base, y charge le jeu de démonstration, puis ouvre le guichet."""
    if not DOMAINE:
        raise RuntimeError("LAFIA_DOMAINE manquant : domaine de base des applications.")
    base = await BaseDIdentite.ouvrir()
    await demonstration.charger(base)
    app.state.guichet = Guichet(base, signataire)
    yield
    await base.fermer()


def guichet_ouvert(request: Request) -> Guichet:
    """Dépendance : le guichet, ouvert au démarrage."""
    guichet: Guichet = request.app.state.guichet
    return guichet


def application_de_la_requete(request: Request) -> Application:
    """Dépendance : l'application de l'hôte de la requête. 404 quand l'hôte n'en désigne aucune."""
    application = application_de_l_hote(request.headers.get("host", ""), DOMAINE)
    if application is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="aucune application à cette adresse")
    return application


def formulaire_de_l_application(
    request: Request, application: Application = Depends(application_de_la_requete)
) -> Application:
    """Dépendance d'un formulaire : l'application, s'il vient de ses propres pages. 403 sinon, pour
    qu'une page hostile ne connecte personne sous le compte d'un autre."""
    if request.headers.get("origin") != origine(application, DOMAINE):
        raise HTTPException(status.HTTP_403_FORBIDDEN, detail="formulaire venu d'une autre origine")
    return application


def adresse_source(request: Request) -> str:
    """L'adresse du client, que la passerelle transmet dans X-Forwarded-For. Caddy remplace celle
    qu'enverrait le client : il ne fait confiance à aucun mandataire en amont."""
    transmise = request.headers.get("x-forwarded-for", "").rsplit(",", 1)[-1].strip()
    return transmise or (request.client.host if request.client else "inconnue")


def _poser_les_cookies(reponse: Response, session: SessionOuverte) -> None:
    for nom, valeur, duree in (
        (COOKIE_DE_SESSION, session.jeton, session.duree_du_jeton),
        (COOKIE_DE_RENOUVELLEMENT, session.identifiant, session.duree_restante),
    ):
        reponse.set_cookie(
            nom, valeur, max_age=int(duree.total_seconds()), path="/", secure=True, httponly=True, samesite="strict"
        )


def _effacer_les_cookies(reponse: Response) -> None:
    for nom in (COOKIE_DE_SESSION, COOKIE_DE_RENOUVELLEMENT):
        reponse.delete_cookie(nom, path="/", secure=True, httponly=True, samesite="strict")


def _issue_de_la_connexion(resultat: SessionOuverte | Refus) -> RedirectResponse:
    """Vers l'accueil avec les deux cookies ; ou vers la page de connexion, avec la raison du refus."""
    if isinstance(resultat, Refus):
        return RedirectResponse(f"/connexion?erreur={resultat}", status_code=status.HTTP_303_SEE_OTHER)
    reponse = RedirectResponse("/", status_code=status.HTTP_303_SEE_OTHER)
    _poser_les_cookies(reponse, resultat)
    return reponse


REDIRECTIONS_DE_CONNEXION: dict[int | str, dict[str, object]] = {
    303: {"description": "Vers `/` avec les deux cookies ; ou vers `/connexion?erreur=identifiants|application|verrouille`."},
    403: {"description": "Formulaire venu d'une autre origine que l'application."},
}

routes = APIRouter()


@routes.post("/connexion", status_code=status.HTTP_303_SEE_OTHER, responses=REDIRECTIONS_DE_CONNEXION)
async def connexion(
    identifiant: Annotated[str, Form()] = "",
    mot_de_passe: Annotated[str, Form()] = "",
    application: Application = Depends(formulaire_de_l_application),
    adresse: str = Depends(adresse_source),
    guichet: Guichet = Depends(guichet_ouvert),
) -> RedirectResponse:
    """Connecte un agent ou une officine sur l'application dont vient le formulaire."""
    return _issue_de_la_connexion(await guichet.connecter(identifiant, mot_de_passe, application, adresse))


@routes.post("/citoyen/connexion", status_code=status.HTTP_303_SEE_OTHER, responses=REDIRECTIONS_DE_CONNEXION)
async def connexion_citoyen(
    npi: Annotated[str, Form()] = "",
    code: Annotated[str, Form()] = "",
    application: Application = Depends(formulaire_de_l_application),
    adresse: str = Depends(adresse_source),
    guichet: Guichet = Depends(guichet_ouvert),
) -> RedirectResponse:
    """Connecte un citoyen par son NPI et son code carnet, sur l'application citoyen seulement. Un NPI
    inconnu et un mauvais code reçoivent la même réponse."""
    return _issue_de_la_connexion(await guichet.connecter_citoyen(npi, code, application, adresse))


@routes.post(
    "/session/renouveler",
    status_code=status.HTTP_204_NO_CONTENT,
    responses={401: {"description": "Session finie, fermée ou inconnue : les deux cookies sont effacés."}},
)
async def renouveler(
    application: Application = Depends(application_de_la_requete),
    identifiant_de_session: Annotated[str | None, Cookie(alias=COOKIE_DE_RENOUVELLEMENT)] = None,
    guichet: Guichet = Depends(guichet_ouvert),
) -> Response:
    """Un nouveau jeton, depuis le cookie de renouvellement. L'application l'appelle avant de rendre une
    page quand le jeton a expiré ; un renouvellement compte comme une activité."""
    session = await guichet.renouveler(identifiant_de_session, application) if identifiant_de_session else None
    if session is None:
        refus = JSONResponse({"detail": "session terminée"}, status_code=status.HTTP_401_UNAUTHORIZED)
        _effacer_les_cookies(refus)
        return refus
    reponse = Response(status_code=status.HTTP_204_NO_CONTENT)
    _poser_les_cookies(reponse, session)
    return reponse


@routes.post(
    "/deconnexion",
    status_code=status.HTTP_303_SEE_OTHER,
    responses={
        303: {"description": "Vers `/connexion`, les deux cookies effacés."},
        403: {"description": "Formulaire venu d'une autre origine que l'application."},
    },
)
async def deconnexion(
    _application: Application = Depends(formulaire_de_l_application),
    identifiant_de_session: Annotated[str | None, Cookie(alias=COOKIE_DE_RENOUVELLEMENT)] = None,
    guichet: Guichet = Depends(guichet_ouvert),
) -> RedirectResponse:
    """Ferme la session : plus aucun jeton n'en sera renouvelé. Le jeton en cours expire de lui-même."""
    if identifiant_de_session:
        await guichet.deconnecter(identifiant_de_session)
    reponse = RedirectResponse("/connexion", status_code=status.HTTP_303_SEE_OTHER)
    _effacer_les_cookies(reponse)
    return reponse


@routes.get(
    "/session",
    responses={401: {"description": "Jeton absent, mal formé, expiré ou mal signé, ou d'un compte inconnu."}},
)
async def session(
    porteur: Porteur = Depends(jetons.porteur), guichet: Guichet = Depends(guichet_ouvert)
) -> SessionAgent | SessionOfficine | SessionCitoyen:
    """Le porteur du jeton, tel que son application l'affiche : un agent, son nom et son établissement ;
    une officine, son nom ; un citoyen, son seul identifiant, jamais son NPI."""
    presente = await guichet.presenter(porteur)
    if presente is None:
        raise HTTPException(status.HTTP_401_UNAUTHORIZED, detail="aucun compte pour ce jeton")
    return presente


class DemandeDeCode(BaseModel):
    npi: Annotated[str, StringConstraints(pattern=rf"^{NPI.pattern}$")]


class CodeCarnet(BaseModel):
    code: str
    """À imprimer sur le reçu : identite n'en garde que l'empreinte, et ne le montre plus jamais."""


@routes.post(
    "/codes-carnet",
    status_code=status.HTTP_201_CREATED,
    responses={
        401: {"description": "Jeton absent, mal formé, expiré ou mal signé."},
        403: {"description": "Un autre rôle que médecin ou infirmier."},
        422: {"description": "NPI mal formé : treize chiffres."},
    },
)
async def codes_carnet(
    demande: DemandeDeCode, soignant: Agent = Depends(soignant_connecte), guichet: Guichet = Depends(guichet_ouvert)
) -> CodeCarnet:
    """Un nouveau code carnet pour le reçu du patient : il remplace le précédent de ce NPI (ADR 0005)."""
    return CodeCarnet(code=await guichet.emettre_code(demande.npi, soignant))


@routes.get("/comptes-de-demonstration")
async def comptes_de_demonstration(
    application: Application = Depends(application_de_la_requete),
) -> list[CompteDeDemonstration | CitoyenDeDemonstration]:
    """Les comptes de démonstration des rôles que l'application connecte, et leurs mots de passe,
    publics : tout le jeu est synthétique. Jamais ceux réservés aux tests."""
    return demonstration.comptes_de_demonstration(application)


app = creer_service(SERVICE, routes, parle_fhir=False, cycle_de_vie=cycle_de_vie)

File: web/commun/src/service.ts
// Le service d'une application, et identite, vus du serveur de l'application. Ils ne sont joints que
// par la passerelle, à l'entrée interne du sous-domaine de l'application : l'application n'est pas sur
// le réseau du noyau et ne parle jamais FHIR.

export const DELAI_MS = 5000;

/** Cookie du jeton de chaque application, celui que les services lisent (`commun.jeton.COOKIE_DE_SESSION`). */
export const COOKIE_DE_SESSION = "__Host-session";

/** Cookie de la session qui renouvelle le jeton ; seul identite le lit. */
export const COOKIE_DE_RENOUVELLEMENT = "__Host-renouvellement";

/** Réponse de `GET /api/<service>/sante` d'un service qui parle FHIR (`commun.service.Sante`). */
export type Sante = {
  service: string;
  statut: "disponible" | "indisponible";
  noyau: {
    statut: "disponible" | "injoignable";
    version_fhir: string | null;
  };
};

/** Le rôle d'un agent, professionnel rattaché à un établissement (`commun.jeton.Role`). */
export type RoleAgent = "médecin" | "infirmier" | "caissier" | "pharmacien";

/** Un agent connecté, tel que `GET /api/identite/session` le rend : de quoi le nommer, lui et son établissement. */
export type SessionAgent = {
  sub: string;
  role: RoleAgent;
  etablissement: string;
  nom: string;
  nom_etablissement: string;
};

/** Une officine connectée : son nom ; elle n'est pas un établissement. */
export type SessionOfficine = {
  sub: string;
  role: "officine";
  nom: string;
};

/** Le citoyen connecté, sans son NPI, que ni identite ni son service ne renvoient jamais. */
export type SessionCitoyen = {
  sub: string;
  role: "citoyen";
};

/** Qui est connecté, selon identite : un agent, une officine ou le citoyen. Le rôle les distingue. */
export type Session = SessionAgent | SessionOfficine | SessionCitoyen;

/** Un compte de démonstration d'agent ou d'officine, que la page de connexion liste (mot de passe public). */
export type CompteDeDemonstration = {
  identifiant: string;
  mot_de_passe: string;
  role: RoleAgent | "officine";
  /** Le nom de l'établissement de l'agent, ou celui de l'officine. */
  structure: string;
};

/** Un citoyen de démonstration, que la page de connexion du citoyen liste. */
export type CitoyenDeDemonstration = {
  npi: string;
  code: string;
  role: "citoyen";
};

/** Adresse interne de la passerelle pour le sous-domaine de l'application, lue dans `PASSERELLE_URL`. */
export function adressePasserelle(): string {
  const adresse = process.env.PASSERELLE_URL;
  if (!adresse) {
    throw new Error("PASSERELLE_URL manquante : adresse interne de la passerelle pour cette application.");
  }
  return adresse;
}

/**
 * Demande `chemin` par la passerelle. Rend le corps JSON quand le service répond par l'un des
 * `statutsAttendus` ; `null` pour tout autre statut, ou quand il ne répond pas.
 */
async function demander<T>(
  service: string,
  chemin: string,
  statutsAttendus: number[],
  enTetes: HeadersInit = {},
): Promise<T | null> {
  const adresse = `${adressePasserelle()}${chemin}`;
  try {
    const reponse = await fetch(adresse, {
      cache: "no-store",
      headers: enTetes,
      signal: AbortSignal.timeout(DELAI_MS),
    });
    if (!statutsAttendus.includes(reponse.status)) {
      console.warn(`service ${service} : ${adresse} a répondu ${reponse.status}`);
      return null;
    }
    return (await reponse.json()) as T;
  } catch (erreur) {
    console.warn(`service ${service} injoignable : ${adresse}`, erreur);
    return null;
  }
}

/** Ce qu'une application demande à son service. */
export type Service = {
  /** État du service et du noyau derrière lui ; `null` quand le service ne répond pas. */
  lireSante(): Promise<Sante | null>;
};

/** Le service `service` (soin, caisse, …), joint sous `/api/<service>` par la passerelle. */
export function serviceParLaPasserelle(service: string): Service {
  return {
    // 200 : toute la chaîne répond. 503 : le service répond, le noyau non. Les deux décrivent l'état.
    lireSante: () => demander<Sante>(service, `/api/${service}/sante`, [200, 503]),
  };
}

/** Ce qu'une application demande à identite. */
export type Identite = {
  /**
   * Le porteur du jeton de session, vérifié par identite : l'application ne lit pas le jeton
   * elle-même. `null` quand identite le refuse ou ne répond pas.
   */
  lireSession(jeton: string): Promise<Session | null>;
  /** Les comptes de démonstration des rôles que l'application connecte ; `null` quand identite ne répond pas. */
  lireComptesDeDemonstration(): Promise<(CompteDeDemonstration | CitoyenDeDemonstration)[] | null>;
};

/** identite, joint sous `/api/identite` par la passerelle, sur le sous-domaine de l'application. */
export function identiteParLaPasserelle(): Identite {
  return {
    lireSession: (jeton) =>
      demander<Session>("identite", "/api/identite/session", [200], {
        Cookie: `${COOKIE_DE_SESSION}=${jeton}`,
      }),
    lireComptesDeDemonstration: () =>
      demander<(CompteDeDemonstration | CitoyenDeDemonstration)[]>(
        "identite",
        "/api/identite/comptes-de-demonstration",
        [200],
      ),
  };
}
